Min Web API tutorial rewrite plus app samples updates v10 (#37294)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/min-web-api.md

Reviewer: If the app is restarted, GET for the list of todos returns an empty array, and GET with an id field will return a `404 Not Found`.

Author: Addressed in latest commit.  Thanks.

Reviewer: ```suggestion
Use the Swagger UI to send a PUT request:
```

Reviewer: ```suggestion
Use the Swagger UI to send a PATCH request:
```

Reviewer: ```suggestion
Use the Swagger UI to send a DELETE request:
```

* Start Visual Studio 2026 and select **Create a new project**.
* In the **Create a new project** dialog:
  * Select the **ASP.NET Core Web API** project type, and select **Next**.
  * Name the project *TodoApi*, and select **Next**.
* In the **Additional information** dialog:
   * Confirm the **Framework** is **.NET 10.0 (Long Term Support)**.
   * Confirm the checkbox for **Enable OpenAPI support** is checked.
   * Confirm the checkbox for **Use controllers** is **not** checked. Uncheck this setting to create a Minimal API project as required for this tutorial rather than a controller-based one.
   * Select **Create**.

  ![Additional information](~/tutorials/min-web-api/static/10/add-information-dialog-18-7-1.png)

# [Visual Studio Code](#tab/visual-studio-code)

* Start Visual Studio Code, select **View**, and then select **Terminal** to open the [integrated terminal](https://code.visualstudio.com/docs/editor/integrated-terminal).

> * Change directories (`cd`) to the folder that contains the project folder.

Reviewer 2: ```suggestion
* Change directory(`cd`) to the folder that contains the project folder.
```

## Replace the WeatherForecast sample with the Todo API

The `webapi` template added a sample `GET /weatherforecast` endpoint to `Program.cs` and a `WeatherForecast` record at the bottom of the file. The sample is just a placeholder to show the template works — replace both with the Todo endpoints below.

* In `Program.cs`, replace the `WeatherForecast` sample endpoint and the `WeatherForecast` record with the following code:

 :::code language="csharp" source="~/tutorials/min-web-api/samples/10.x/todo/Program.cs?name=snippet_minimal_start_all":::

The OpenAPI lines added by the template (`builder.Services.AddOpenApi()` and `app.MapOpenApi()`) remain in place. They now describe the Todo endpoints instead of the sample weather endpoint.

The following highlighted code adds the database context to the [dependency injection (DI)](xref:fundamentals/dependency-injection) container and enables displaying database-related exceptions:

[!code-csharp[](~/tutorials/min-web-api/samples/10.x/todo/Program.cs?name=snippet_DI&highlight=2)]

> The DI container provides access to the database context and other services.

Reviewer: I think it would be good to explain that "ToDoList" is the name of the in-memory database.

Author: Addressed in latest commit.

# [Visual Studio Code](#tab/visual-studio-code)

### Add a browser UI to view the OpenAPI document

> The template already generates the OpenAPI document at `/openapi/v1.json`. To explore and test the API from a browser, add a UI that consumes that document. This tutorial uses NSwag.

Reviewer: Is there some reason to use NSwag? Our docs:

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/using-openapi-documents?view=aspnetcore-10.0

describe how to add either SwaggerUI (from Swashbuckle) or Scalar but not NSwag. Should we be consistent?

Author: I accidently used it out of old habit I think.  It was a mistake.  I was using SwaggerUI, which is packaged in NSwag but I should have used the Swashbuckle package if doing so.  Thanks.

Author: Changed to Scalar now instead.

# [Visual Studio Code](#tab/visual-studio-code)

> * With the app still running, in the browser, navigate to `https://localhost:{port}/swagger` to display the API testing page generated by Swagger.

Reviewer: ```suggestion
* With the app still running, in the browser, navigate to `https://localhost:{port}/swagger` to display the API testing page generated by NSwag.
```
or other package if we decided to use SwaggerUI or Scalar.

Author: I will include this change so I have a cleaner version, then locally change everything to Scalar, including screenshots.

> * In Swagger select **GET /todoitems** > **Try it out** > **Execute**.

Reviewer: ```suggestion
* In the Swagger UI select **GET /todoitems** > **Try it out** > **Execute**.
```

# [Visual Studio](#tab/visual-studio)

* In **Endpoints Explorer**, right-click the **PUT** endpoint, and select **Generate request**.

  The following content is added to the `TodoApi.http` file:

```http
  PUT {{TodoApi_HostAddress}}/todoitems/{id}

  ###
```

* In the PUT request line, replace `{id}` with `1`.

* Add the following lines immediately after the PUT request line:

```http
  Content-Type: application/json

  {
    "id": 1,
    "name": "feed fish",
    "isComplete": false
  }
```

  The preceding code adds a Content-Type header and a JSON request body.

* Select the **Send request** link that is above the new PUT request line.
  
# [Visual Studio Code](#tab/visual-studio-code)

Use the Scalar UI to send a PUT request:

* Select **PUT /todoitems/{id}**, and then select **Test Request**.

* Set the **id** field to `1`.

* Set the request body to the following JSON:

```json
  {
    "id": 1,
    "name": "feed fish",
    "isComplete": false
  }
```

* Select **Send**.

---

The PUT request is sent to the app and the response is displayed in the **Response** pane. The response body is empty, and the status code is 204.

## Create and examine the PATCH endpoint

> A PATCH endpoint lets clients send only the fields they want to update, such as renaming a Todo item without resending its completion status. This approach differs from a PUT request, which replaces the entire item, so the client must send every field even to change just one.

Reviewer 2: ```suggestion
A PATCH endpoint lets clients send only the fields they want to update, such as renaming a Todo item without resending its completion status. This approach differs from a PUT request, which replaces the entire item, so the client must send every field even if only one is changed.
```
